sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala

```diff
@@ -675,12 +675,23 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
     val useAdvanced = SQLConf.get.advancedPartitionPredicatePushdownEnabled
 
     object ExtractAttribute {
+      val partitionKeys = table.getPartitionKeys.asScala.map(_.getName).toSet
+      var castToStr = false
+
       def unapply(expr: Expression): Option[Attribute] = {
         expr match {
-          case attr: Attribute => Some(attr)
+          case attr: Attribute
```

Comment:
can't we simply do `case attr: Attribute if attr.dataType == StringType`?

Reply:
I used to test that, but it can't pass some tests in `PartitionedTablePerfStatsSuite`.
It seems that such as `p1 = '0'` (p1 is a Int partition key and  '0' is an Integer) can filter some partitions, but `p1 = "0"` (p1 is a Int partition key and "0" is a String) can't be pushed down.
So, I add the prediction here to judge whether it has castToString operation.

Reply:
I will remove the `!partitionKeys.contains(attr.name)`,  it is false always. 
I did't have a good understanding about this.

Reply:
can you point to the problematic test cases?

Reply:
Such as :
```
genericTest("lazy partition pruning reads only necessary partition data")
```
The relative log is 
```
5 did not equal 0
ScalaTestFailureLocation: org.apache.spark.sql.hive.PartitionedTablePerfStatsSuite at (PartitionedTablePerfStatsSuite.scala:139)
Expected :0
Actual   :5
<Click to see difference>

org.scalatest.exceptions.TestFailedException: 5 did not equal 0
```

Reply:
Now I'm a little confused. Seems hive does support to filter non-string-type partition columns.

Reply:
seems we need to revisit https://github.com/apache/spark/pull/19602

Reply:
But if you execute a sql like 
```
sql("SELECT c1 FROM t1 WHERE CAST(p1 as STRING) = '5'").show
```
It will throw a exception:
```
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.sql.hive.client.Shim_v0_13.getPartitionsByFilter(HiveShim.scala:759)
	... 57 more
Caused by: MetaException(message:Filtering is supported only on partition keys of type string)
```

Reply:
ok.